### ISSUE TITLE
Proper handling of contact record creation when account is publicly registered?

### ISSUE THREAD

**Issue by dnotestein** [BitShares/bitshares#1283]
If a user attempts to create a contact record using an account name that is already publicly registered, the client rejects the contact record with the message: 

Invalid account name
Account name is already registered under a different key.

This would make sense when the key the user was trying to tie the account name to was a different one than the one it is registered under, but this error is displayed even if the keys match. At the very least, we need to change the error message, but I think it would be better to compare the registered key and the key the user is trying to tie to the account and allow the contact record to be added if they match. There's some evidence in my wallet that things worked this way in the past, so it may have been inadvertently changed to the current problematic behavior (or maybe it's some key formatting weirdness, for that matter; seems we have a lot of that).

Another potential problem area is what happens when a user creates a local contact record for a name, then someone else later registers that name publicly with a different key. Presumably the user's wallet will still use the key in the local contact record instead of the registered name. In other words, the local binding for the name should always have precedence over the registered one (hopefully this is already the case).

**Comment by testzcrypto** [comment]
To solve this problem, I thinks we should add the prefix to any local contacts created by user (let's say local-yyyyyy) and not allow to create names with this prefix in blockchain.

**Comment by vikramrajkumar** [comment]
A general thought: I've recently been feeling like we should completely separate the notion of accounts, contacts, and voting slate in the wallet--rather than treating them all as members of the same `wallet_account_record` list.

> Another potential problem area is what happens when a user creates a local contact record for a name, then someone else later registers that name publicly with a different key. Presumably the user's wallet will still use the key in the local contact record instead of the registered name. In other words, the local binding for the name should always have precedence over the registered one (hopefully this is already the case).

In this case I believe the command is supposed to fail until you rename the contact to something that doesn't conflict, but haven't tested this in a while.

**Comment by dnotestein** [comment]
>In this case I believe the command is supposed to fail until you rename the contact to something that doesn't conflict, but haven't tested this in a while.

This doesn't seem like a good idea to me. For example, what if I have a script written that periodically makes a transfer to a local account. My script will start failing as soon as someone registers that name publicly. This would basically force such users to use public keys rather than local account names to prevent this from happening (assuming they knew about this behavior, which seems doubtful).

**Comment by vikramrajkumar** [comment]
What if we separated contacts and accounts as I mentioned above, and then make it so that you can only transfer to public keys, addresses, or contacts. And a contact is just a label for a registered account name, public key, or address.

Basically I've been thinking that we should eliminate the notion of "unregistered accounts", which seems to be a misnomer that causes more trouble than it is worth.

**Comment by dnotestein** [comment]
Ok, so the idea would be to allow users to assign an "alias" (label) to any potential key whether it's registered or not, then the user can reference that alias in his api calls? That sounds fine to me as far as it goes, as long as we haven't built up too many assumptions in the client code about the nature of unregistered account names. 

Essentially adding a contact would be to add an alias, correct? So we would modify the wallet_add_contact_account api call to allow a registered account instead of the normally accepted key (and maybe create an alias for the api call that's more accurately named for the new functionality). Presumably, local aliases would always take precedence over account names when referenced in api calls. So if someone makes an api call referencing an alias account name to a publicly registered account, we would substitute the "real" account name for the alias when processing the command  (i.e. we could "preprocess" replace local aliases with their mapping). I think this is necessary to avoid problems if someone maps an alias to a publicly registered account, then the active key (or whatever we're calling it nowadays) is changed to some other key. Will we allow multiple aliases to the same account or key?

For display of unregistered accounts, will we show an alias if there is one? Do we require an initial alias to be specified for unregistered accounts? If someone goes to publicly register an account for the first time, do we "suggest" the alias if one is specified at the GUI level as the "registered" name, similar to what we do now?

Basically, I think your idea is sound, but if we want to implement this, I think we need to hammer down these details since it's a significant change in user experience.

**Comment by theoreticalbts** [comment]
What if the wallet API *only* lets you send to contact accounts?  So adding someone as a contact account is necessary in order to send them funds.  In the GUI the dropdown / autocomplete should only choose from contacts, but the user will have a streamlined way to add new contacts directly from the transfer UI without going to a separate contacts page.  E.g. if the user types a name that isn't a contact, it will give the user a dialog:  "This person is not in your contacts.  Add registered account `<name>` as contact?  `<robohash>`  Notes:  <text field for entering locally stored notes about this contact>  [Add as contact] [Cancel]"

The additional dialog will be an extra layer of defense against typos.  But it won't be too inconveniencing since user will only need to deal with it once per contact.

With this system, if you wanted to send to a key or a TITAN account, you'd have to add them as a contact account first.  Then the contact account becomes the local alias.  N.b. a contact that represents a TITAN account should only point to the TITAN account name, and should get the address associated with that name from the blockchain anew for each transfer.  So if the TITAN account changes its key after being added, the contact will send funds to the updated address.

**Comment by vikramrajkumar** [comment]
> What if the wallet API only lets you send to contact accounts?

I agree this would be a good default, we may want to allow bypassing it however for e.g. exchanges.

> N.b. a contact that represents a TITAN account should only point to the TITAN account name, and should get the address associated with that name from the blockchain anew for each transfer.

Yes.

I am starting to envision an ontology something like:
- accounts
  - each is just a pointer to an account id for which you control a non-empty subset of {owner, active, signing} keys
- receive keys
  - each is just a pointer to a public key for which you control the private key
- contacts
  - each is just a pointer to a member of {account id, public key, address}

And each of the above can have any custom label, all of which would exist in a separate label namespace.

All accounts start out as receive keys, which are then "upgraded" to accounts by paying the network for registration.

**Comment by vikramrajkumar** [comment]
Superseded by: #1401
